plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java

```diff
@@ -0,0 +1,251 @@
+    final Map<String, Object> distTags = (Map<String, Object>) rootJson.get("dist-tags");
+    distTags.remove(lastVersion);
+    distTags.put("latest", lastVersion);
+
+    final PackageRoot packageRoot = new PackageRoot(context.getRepository().getId(), rootJson);
+    hostedMetadataService.consumePackageRoot(packageRoot);
+    packageRoots = packageRoots + 1;
+    packageVersions = packageVersions + versions.size();
+  }
+
+  private Map<String, Object> extractPackageJson(final StorageFileItem file) throws IOException {
+    try (final GZIPInputStream gzipInputStream = new GZIPInputStream(file.getInputStream())) {
+      log.debug("Examining TAR file");
+      final TarArchiveInputStream tarArchiveInputStream = new TarArchiveInputStream(gzipInputStream);
+      TarArchiveEntry tarEntry;
+      // npm uses tar option "--strip-components=1", so we need to strip first path element to have proper names
+      String tarEntryName;
+      do {
+        tarEntry = tarArchiveInputStream.getNextTarEntry();
+        tarEntryName = null;
+        if (tarEntry != null) {
+          // tar entry names are "pathelem1/pathelem2/.../filename"
+          int indexOfSlash = tarEntry.getName().indexOf("/");
+          if (indexOfSlash > 0) {
+            tarEntryName = tarEntry.getName().substring(indexOfSlash);
+          }
+        }
+        log.debug("tarEntry={}, name={}", tarEntry.getName(), tarEntryName);
+      }
+      while (tarEntryName != null && !PACKAGE_JSON_PATH.equals(tarEntryName));
+      if (tarEntryName == null) {
+        return null;
+      }
+      // checks for corrupted data, we do want to report these
+      checkArgument(tarEntry.isFile(), "Tar content %s not a file?", PACKAGE_JSON_PATH);
+      checkArgument(tarEntry.getSize() < MAX_PACKAGE_JSON_FILE_SIZE, "Tar content too big %s: %s bytes",
+          PACKAGE_JSON_PATH, tarEntry.getSize());
+      // do this in memory, as package.json is expected to be fairly small (few kb at most)
+      final ByteArrayOutputStream bos = new ByteArrayOutputStream(Ints.checkedCast(tarEntry.getSize()));
+      StreamSupport.copy(tarArchiveInputStream, bos, StreamSupport.BUFFER_SIZE);
+      final Map<String, Object> packageJson = objectMapper
+          .readValue(bos.toByteArray(), new TypeReference<HashMap<String, Object>>() { });
+      return packageJson;
+    }
+  }
+
+  private void processPackageJson(final Map<String, Object> rootJson, final Map<String, Object> packageJson)
+      throws Exception
+  {
+    liftAttributes(rootJson, packageJson);
+    final String version = String.valueOf(packageJson.get("version"));
+    final Map<String, Object> distTags = (Map<String, Object>) rootJson.get("dist-tags");
+    final Map<String, Object> versions = (Map<String, Object>) rootJson.get("versions");
+    distTags.put(version, version);
+    versions.put(version, packageJson);
+  }
+
+  private void liftAttributes(final Map<String, Object> higher, final Map<String, Object> lower) {
```

Comment:
input parameter `packageJson` is being added to `rootJson`. So, it is "lift" one level upper (from child packageJson to parent rootJson). But I agree, am known to be terrible at naming things :)